src/metabase/driver.clj

```diff
@@ -330,6 +330,14 @@
   dispatch-on-initialized-driver
   :hierarchy #'hierarchy)
 
+(defmulti execute-query-callback
+  "Like execute-query, but pass the result to a callback rather than returning it."
+  {:arglists '([driver query callback]), :style/indent 1}
+  dispatch-on-initialized-driver
+  :hierarchy #'hierarchy)
+
+(defmethod execute-query-callback :default [driver query callback]
+  (callback (execute-query driver query)))
```

Comment:
This is expected to run for all non-SQL databases, right? I had a hunch, and then confirmed that it doesn't work with MongoDB (see exception below).

I expect the same is true for other non-SQL database, although I haven't confirmed that.

I also haven't dug into _why_ this happened. @camsaul do you have any ideas?

```
01-14 00:08:54 DEBUG middleware.log :: POST /api/dataset 200 [ASYNC: completed] 113.0 ms (16 DB calls) Jetty threads: 4/50 (2 idle, 0 queued) (63 total active threads) Queries in flight: 0
01-14 00:09:02 WARN middleware.process-userland-query :: Query failure {:status :failed,
 :class java.lang.Exception,
 :error
 "Query processor error: mismatched number of columns in query and results. Expected 11 fields, got 0\nExpected: [\"_id\" \"id\" \"name\" \"adminregion\" \"capitalCity\" \"incomeLevel\" \"iso2Code\" \"latitude\" \"lendingType\" \"long
itude\" \"region\"]\nActual: []",
 :stacktrace
 ("--> query_processor.middleware.annotate$check_correct_number_of_columns_returned.invokeStatic(annotate.clj:371)"
  "query_processor.middleware.annotate$check_correct_number_of_columns_returned.invoke(annotate.clj:364)"
  "query_processor.middleware.annotate$eval56789$fn__56791.invoke(annotate.clj:437)"
  "query_processor.middleware.annotate$eval56830$add_column_info_STAR___56835$fn__56839.invoke(annotate.clj:475)"
  "query_processor.middleware.annotate$eval56830$add_column_info_STAR___56835.invoke(annotate.clj:470)"
  "query_processor.middleware.annotate$add_column_info$fn__56852.invoke(annotate.clj:485)"
  "query_processor.middleware.format_rows$format_rows$fn__67384.invoke(format_rows.clj:82)"
  "query_processor$eval71850$execute_query__71855$fn__71856$fn__71857.invoke(query_processor.clj:74)"
  "driver$eval36832$fn__36833.invoke(driver.clj:340)"
  "query_processor$eval71850$execute_query__71855$fn__71856.invoke(query_processor.clj:71)"
  "query_processor$eval71850$execute_query__71855.invoke(query_processor.clj:64)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__64631.invoke(mbql_to_native.clj:40)"
  "query_processor.middleware.annotate$result_rows_maps__GT_vectors$fn__56946.invoke(annotate.clj:541)"
  "query_processor.middleware.annotate$add_column_info$fn__56852.invoke(annotate.clj:485)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__66722.invoke(cumulative_aggregations.clj:57)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__68529.invoke(resolve_joins.clj:184)"
  "query_processor.middleware.limit$limit$fn__67404.invoke(limit.clj:19)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__71650.invoke(results_metadata.clj:87)"
  "query_processor.middleware.format_rows$format_rows$fn__67384.invoke(format_rows.clj:82)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__65051.invoke(add_dimension_projections.clj:234)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__65718.invoke(add_source_metadata.clj:107)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__68587.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__65563.invoke(add_row_count_and_status.clj:16)"
  "driver.mongo$eval101959$fn__101960$fn__101962$f__100870__auto____101964.invoke(mongo.clj:86)"
  "driver.mongo.util$_with_mongo_connection$fn__100861.invoke(util.clj:205)"
  "util.ssh$with_ssh_tunnel_STAR_.invokeStatic(ssh.clj:106)"
  "util.ssh$with_ssh_tunnel_STAR_.invoke(ssh.clj:91)"
  "driver.mongo.util$_with_mongo_connection.invokeStatic(util.clj:199)"
  "driver.mongo.util$_with_mongo_connection.invoke(util.clj:194)"
  "driver.mongo$eval101959$fn__101960$fn__101962.invoke(mongo.clj:85)"
  "query_processor.middleware.driver_specific$process_query_in_context$fn__66819.invoke(driver_specific.clj:12)"
  "query_processor.middleware.resolve_driver$resolve_driver$fn__68177.invoke(resolve_driver.clj:22)"
  "query_processor.middleware.store$initialize_store$fn__71689$fn__71690.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__71689.invoke(store.clj:10)"
  "query_processor.middleware.async$async__GT_sync$fn__64538.invoke(async.clj:23)"
  "query_processor.middleware.async_wait$runnable$fn__65785.invoke(async_wait.clj:89)"),
 :query
 {:type "query",
  :query {:source-table 76},
  :parameters [],
  :async? true,
  :middleware {:userland-query? true, :skip-results-metadata? true},
  :data-fn #function[metabase.api.dataset/eval72146/as-format-async--72151/fn--72152/fn--72154],
  :info
  {:executed-by 1,
   :context :csv-download,
   :query-hash [33, 103, -29, -79, -26, -51, 119, -34, 115, -111, 91, 44, -20, -90, -36, 54, -113, 63, 39, -64, -100, -93, 70, 101, 23, 12, -66, 73, 101, -28, 29, 124]}},
 :preprocessed
 {:type :query,
  :middleware {:userland-query? true, :skip-results-metadata? true},
  :data-fn #function[metabase.api.dataset/eval72146/as-format-async--72151/fn--72152/fn--72154],
  :info
  {:executed-by 1,
   :context :csv-download,
   :query-hash [33, 103, -29, -79, -26, -51, 119, -34, 115, -111, 91, 44, -20, -90, -36, 54, -113, 63, 39, -64, -100, -93, 70, 101, 23, 12, -66, 73, 101, -28, 29, 124]},
  :preprocessing-level 1,
  :database 27,
  :query
  {:source-table 76,
   :fields [[:field-id 134] [:field-id 135] [:field-id 138] [:field-id 141] [:field-id 140] [:field-id 142] [:field-id 133] [:field-id 137] [:field-id 143] [:field-id 136] [:field-id 139]],
   :limit 1048576}},
 :native
 {:projections (:_id :id :name :adminregion :capitalCity :incomeLevel :iso2Code :latitude :lendingType :longitude :region),
  :query
  [{"$project"
    {"_id" "$_id",
     "id" "$id",
     "name" "$name",
     "adminregion" "$adminregion",
     "capitalCity" "$capitalCity",
     "incomeLevel" "$incomeLevel",
     "iso2Code" "$iso2Code",
     "latitude" "$latitude",
     "lendingType" "$lendingType",
     "longitude" "$longitude",
     "region" "$region"}}
   {"$project"
    {"_id" "$_id",
     "id" "$id",
     "name" "$name",
     "adminregion" "$adminregion",
     "capitalCity" "$capitalCity",
     "incomeLevel" "$incomeLevel",
     "iso2Code" "$iso2Code",
     "latitude" "$latitude",
     "lendingType" "$lendingType",
     "longitude" "$longitude",
     "region" "$region"}}
   {"$limit" 1048576}],
  :collection "countries",
  :mbql? true}}

01-14 00:09:02 DEBUG middleware.log :: POST /api/dataset/csv 400 259.1 ms (35 DB calls)
"Query processor error: mismatched number of columns in query and results. Expected 11 fields, got 0\nExpected: [\"_id\" \"id\" \"name\" \"adminregion\" \"capitalCity\" \"incomeLevel\" \"iso2Code\" \"latitude\" \"lendingType\" \"longi
tude\" \"region\"]\nActual: []"
```

Test data imported with:

```
curl https://api.worldbank.org/countries\?per_page\=304\&format\=json | jq '.[1]' > countries.json
mongoimport -d metabase_11554_test -c countries --jsonArray < countries.json
```

Reply:
This is directly related to the ad hoc middleware invocation discussed in https://github.com/metabase/metabase/pull/11554#discussion_r366109790. Adding the `annotate/result-rows-maps->vectors` middleware fixes it.